### ISSUE TITLE
Allow configuration of spot instance eviction policy

### ISSUE THREAD

**Issue by tchrischan** [jenkinsci/azure-vm-agents-plugin#322]
### Feature Request

Thanks for supporting spot VMs with this plugin. Can you please allow configuration of the eviction policy for spot VMs? We would like to be able to set a maximum price so VMs aren't evicted in the middle of a build.

**Comment by timja** [comment]
Isn’t it set to -1 which is the same as regular pricing?

No expert and not much experience in this

**Comment by tchrischan** [comment]
@timja I would also expect, even welcome, `--spot-max-price -1`, but apparently it does not.

**Comment by tchrischan** [comment]
I also opened https://github.com/jenkinsci/azure-vm-agents-plugin/issues/323 for a related side effect, but the current eviction policy is stop/deallocate and is not configurable in the Jenkins dialog.

**Comment by timja** [comment]
https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/2019-03-01/virtualmachines?tabs=bicep#billingprofile

The default according to docs is `-1`

**Comment by tchrischan** [comment]
I just launched one of these right now. Jenkins 2.303.2, plugin ver. 797.v31f530348574:
<img width="346" alt="Screen Shot 2021-11-04 at 2 04 38 PM" src="https://user-images.githubusercontent.com/31424159/140422165-f52b88f5-f769-44a0-ba78-b7ffd518c339.png">

**Comment by tchrischan** [comment]
https://github.com/jenkinsci/azure-vm-agents-plugin/blob/d5903dae113991da04ef994cb88dc17b21a30e0a/src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java#L632

**Comment by timja** [comment]
Yes that's right, but the max price is -1 which means it won't get evicted for price reasons but could be for other reasons.

I don't think allowing configuration of the eviction policy  is going to help much here, it will need some code to handle the disconnection and probably abort and retry the build unless anyone has better ideas

**Comment by tchrischan** [comment]
@timja yes I was just reading the link you provided and you are probably correct, perhaps configuration is not the correct result. However it would be terrific if it would honor the -1 maxPrice instead of deleting upon eviction. I would be very happy if the spot VM stayed up and just was deleted per the idle retention rather than deleting without notification.

Excellent suggestion about handling the disconnect and retrying the build (or stage), because that's what I asked for separately in https://github.com/jenkinsci/azure-vm-agents-plugin/issues/323. 😄 I think it would be preferable to deallocate rather than delete by default so the retried build/stage has access to the workspace.

Basically we'd like to use spot for builds but they hang without failing so we have to use non-spot VMs.

**Comment by timja** [comment]
As far as I know it is honouring the max price you're setting, there's two reasons you can get evicted:
https://docs.microsoft.com/en-us/azure/virtual-machines/spot-vms#:~:text=30%20seconds%20notice.-,Eviction%20policy,you%20to%20redeploy%20it%20later.
